File: TheaterSuggestions/CSharp/ScreeningSeatingArrangementSuggestions/AuditoriumSeatingArrangements.cs
using ExternalDependencies.AuditoriumLayoutRepository;
using ExternalDependencies.ReservationsProvider;
using SeatsSuggestions;

public class AuditoriumSeatingArrangements
{
    private readonly AuditoriumLayoutRepository _auditoriumLayoutRepository;
    private readonly ReservationsProvider _reservedSeatsRepository;

    public AuditoriumSeatingArrangements(AuditoriumLayoutRepository auditoriumLayoutRepository,
        ReservationsProvider reservationsProvider)
    {
        _auditoriumLayoutRepository = auditoriumLayoutRepository;
        _reservedSeatsRepository = reservationsProvider;
    }

    public AuditoriumSeatingArrangement FindByShowId(string showId)
    {
        return Adapt(_auditoriumLayoutRepository.GetAuditoriumLayoutFor(showId),
            _reservedSeatsRepository.GetReservedSeats(showId));
    }

    private static AuditoriumSeatingArrangement Adapt(AuditoriumDto auditoriumDto, ReservedSeatsDto reservedSeatsDto)
    {
        var rows = new Dictionary<string, Row>();

        foreach (var rowDto in auditoriumDto.Rows)
        foreach (var seatDto in rowDto.Value)
        {
            var rowName = ExtractRowName(seatDto.Name);
            var number = ExtractNumber(seatDto.Name);
            var pricingCategory = ConvertCategory(seatDto.Category);

            var isReserved = reservedSeatsDto.ReservedSeats.Contains(seatDto.Name);

            if (!rows.ContainsKey(rowName)) rows[rowName] = new Row(rowName, new List<SeatingPlace>());

            rows[rowName].SeatingPlaces.Add(new SeatingPlace(rowName, number, pricingCategory,
                isReserved ? SeatingPlaceAvailability.Reserved : SeatingPlaceAvailability.Available));
        }

        return new AuditoriumSeatingArrangement(rows);
    }

    private static PricingCategory ConvertCategory(int seatDtoCategory)
    {
        return (PricingCategory)seatDtoCategory;
    }

    private static uint ExtractNumber(string name)
    {
        return uint.Parse(name.Substring(1));
    }

    private static string ExtractRowName(string name)
    {
        return name[0].ToString();
    }
}

File: TheaterSuggestions/Java/SeatingPlaceSuggestions/src/main/java/org/weaveit/seatingplacesuggestions/AuditoriumSeatingArrangements.java
package org.weaveit.seatingplacesuggestions;

import org.weaveit.externaldependencies.auditoriumlayoutrepository.AuditoriumDto;
import org.weaveit.externaldependencies.auditoriumlayoutrepository.AuditoriumLayoutRepository;
import org.weaveit.externaldependencies.auditoriumlayoutrepository.SeatDto;
import org.weaveit.externaldependencies.reservationsprovider.ReservationsProvider;
import org.weaveit.externaldependencies.reservationsprovider.ReservedSeatsDto;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AuditoriumSeatingArrangements {

    private final ReservationsProvider reservedSeatsRepository;
    private final AuditoriumLayoutRepository auditoriumLayoutRepository;


    public AuditoriumSeatingArrangements(AuditoriumLayoutRepository auditoriumLayoutRepository, ReservationsProvider reservationsProvider) {
        this.auditoriumLayoutRepository = auditoriumLayoutRepository;
        this.reservedSeatsRepository = reservationsProvider;
    }

    public AuditoriumSeatingArrangement findByShowId(String showId) {
        return adapt(auditoriumLayoutRepository.findByShowId(showId),
                reservedSeatsRepository.getReservedSeats(showId));

    }

    private AuditoriumSeatingArrangement adapt(AuditoriumDto auditoriumDto, ReservedSeatsDto reservedSeatsDto) {

        var rows = new HashMap<String, Row>();

        for (Map.Entry<String, List<SeatDto>> rowDto : auditoriumDto.rows().entrySet()) {
            List<SeatingPlace> seats = new ArrayList<>();

            rowDto.getValue().forEach(seatDto -> {
                var rowName = rowDto.getKey();
                var number = extractNumber(seatDto.name());
                var pricingCategory = convertCategory(seatDto.category());

                var isReserved = reservedSeatsDto.reservedSeats().contains(seatDto.name());

                seats.add(new SeatingPlace(rowName, number, pricingCategory, isReserved ? SeatingPlaceAvailability.RESERVED : SeatingPlaceAvailability.AVAILABLE));
            });

            rows.put(rowDto.getKey(), new Row(rowDto.getKey(), seats));
        }

        return new AuditoriumSeatingArrangement(rows);
    }

    private static PricingCategory convertCategory(int seatDtoCategory) {
        return PricingCategory.valueOf(seatDtoCategory);
    }

    private static int extractNumber(String name) {
        return Integer.parseUnsignedInt(name.substring(1));
    }

}
